cardano-crypto-tests/src/Test/Crypto/Instances.hs

```diff
@@ -1,27 +1,60 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
 {-# OPTIONS_GHC -Wno-orphans #-}
-module Test.Crypto.Instances () where
+module Test.Crypto.Instances
+( withMLSBFromPSB
+, withMLockedSeedFromPSB
+) where
 
 import Data.Maybe (mapMaybe)
 import GHC.Exts (fromListN, toList, fromList)
 import Data.Proxy (Proxy (Proxy))
 import GHC.TypeLits (KnownNat, natVal)
-import Test.QuickCheck (Arbitrary (arbitrary, shrink))
-import qualified Data.ByteString as BS
+import Test.QuickCheck (Arbitrary (..))
+-- import qualified Data.ByteString as BS
 import qualified Test.QuickCheck.Gen as Gen
 import qualified Cardano.Crypto.Libsodium as NaCl
+import Cardano.Crypto.MLockedSeed
 import Cardano.Crypto.PinnedSizedBytes (
   PinnedSizedBytes,
   psbFromByteStringCheck,
   psbToByteString,
   )
+import Test.Crypto.RunIO (RunIO (..))
+import Control.Monad.IO.Class (MonadIO (..))
+import Control.Exception (bracket)
 
-instance KnownNat n => Arbitrary (NaCl.MLockedSizedBytes n) where
-    arbitrary = NaCl.mlsbFromByteString . BS.pack <$> Gen.vectorOf size arbitrary
-      where
-        size :: Int
-        size = fromInteger (natVal (Proxy :: Proxy n))
+-- We cannot allow this instance, because it doesn't guarantee timely
+-- forgetting of the MLocked memory.
+-- Instead, use 'arbitrary' to generate a suitably sized PinnedSizedBytes
+-- value, and then mlsbFromPSB or withMLSBFromPSB to convert it to an
+-- MLockedSizedBytes value.
+--
+-- instance KnownNat n => Arbitrary (NaCl.MLockedSizedBytes n) where
+--     arbitrary = unsafePerformIO . NaCl.mlsbFromByteString . BS.pack <$> vectorOf size arbitrary
+--       where
+--         size :: Int
+--         size = fromInteger (natVal (Proxy :: Proxy n))
```

Comment:
I don't think it makes sense to disallow this instance. Without test suite will suffer and `Arbitrary` is used exclusively in tests.  If someone uses it in production they are just dumb asses and deserve the pain. :smile:

Reply:
The thing is, mlocked memory is a scarce resource (and worse, it's kind of unpredictable how scarce exactly it will be on a given machine); this is one of the reasons why we explicitly allocate and deallocate it in `IO`, and why this `Arbitrary` instance would require `unsafePerformIO`.

QuickCheck defaults to generating 1000 instances of each test, and a Sum5KES key consists of 2^5 subkeys, and we don't have a good way of ensuring timely deallocation of these keys if they are created through this `Arbitrary` + `unsafePerformIO` mechanism - we could easily end up allocating tens of thousands of individual keys before GC decides to deallocate them. We could mess around with triggering GC manually, but frankly I think that's just too messy, and the way that's proposed here, having QuickCheck generate seeds and generating keys from them as needed, inside a proper `IO` context spun up with `ioProperty`, and then explicitly forgetting them the moment we're done with them, is much cleaner, and follows the rule that we always want to be explicit about when mlocked memory is allocated and deallocated, ideally using `with`-style wrappers (although that is not always possible).

I will, however, clean up and reduce the boilerplate a little, so that the tests read more nicely.

Reply:
I have cleaned up a bit more; there are still a few minor warts left that I expect to deal with if/when I move tests to `IOSim`.